sources/osg/StateCache.js

```diff
@@ -0,0 +1,598 @@
+            func: undefined,
+            enable: false,
+            near: undefined,
+            far: undefined
+        },
+        changed: true
+    };
+};
+
+var createStateCullFace = function() {
+    return {
+        buffer: {
+            enable: false,
+            mode: CullFace.BACK
+        },
+        state: {
+            enable: false,
+            mode: undefined
+        },
+        changed: true
+    };
+};
+
+var createStateScissor = function() {
+    return {
+        buffer: {
+            enable: false,
+            x: 0,
+            y: 0,
+            width: 640,
+            height: 480
+        },
+        state: {
+            enable: false,
+            x: undefined,
+            y: undefined,
+            width: undefined,
+            height: undefined
+        },
+        changed: true
+    };
+};
+
+var createStateColorMask = function() {
+    return {
+        buffer: {
+            red: true,
+            green: true,
+            blue: true,
+            alpha: true
+        },
+        state: {
+            red: undefined,
+            green: undefined,
+            blue: undefined,
+            alpha: undefined
+        },
+        changed: true
+    };
+};
+
+var createStateViewport = function() {
+    return {
+        buffer: {
+            x: 0,
+            y: 0,
+            width: 640,
+            height: 480
+        },
+        state: {
+            x: undefined,
+            y: undefined,
+            width: undefined,
+            height: undefined
+        },
+        changed: true
+    };
+};
+
+var createStateDepthMask = function() {
+    return {
+        state: { value: true },
+        buffer: { value: undefined },
+        changed: true
+    };
+};
+
+var createStateClearDepth = function() {
+    return {
+        buffer: { value: 1.0 },
+        state: { value: undefined },
+        changed: true
+    };
+};
+
+var createStateClearColor = function() {
+    return {
+        buffer: {
+            red: 0.0,
+            green: 0.0,
+            blue: 0.0,
+            alpha: 0.0
+        },
+        state: {
+            red: undefined,
+            green: undefined,
+            blue: undefined,
+            alpha: undefined
+        },
+        changed: true
+    };
+};
+
+var StateCache = function() {
+    this._stateClearColor = createStateClearColor();
+    this._stateClearDepth = createStateClearDepth();
+    this._stateDepthMask = createStateDepthMask();
+    this._stateViewport = createStateViewport();
+    this._stateColorMask = createStateColorMask();
+    this._stateScissor = createStateScissor();
+    this._stateDepth = createStateDepth();
+    this._stateCullFace = createStateCullFace();
+    this._stateBlendFunc = createStateBlendFunc();
+};
+
+StateCache.prototype = {
+    clearColor: function(array) {
+        var data = this._stateClearColor;
+        var buffer = data.buffer;
+        var state = data.state;
+        data.changed = false;
+
+        if (
```

Comment:
note that for for multi boolean and/or even four 255 value, you could use bit mask.
One value comparisiong would make it fast.
Attribute could store bitmask and value, as bitmask usually setted not often, but comparison are done a lot of time.
(valable for state without float/integer values)